Move Track Alias from SUBSCRIBE to SUBSCRIBE_OK (#977)
moq-wg/moq-transport: draft-ietf-moq-transport.md

each active upstream subscription that matches that namespace, it SHOULD send a
SUBSCRIBE to the publisher that sent the ANNOUNCE.

Relays use the Track Alias ({{track-alias}}) of an incoming Object to identify

> its track and find the active subscribers for that track. Relays MUST forward

Reviewer: ```suggestion
its track and find the active subscribers. Relays MUST forward
```

  Type (i) = 0x4,
  Length (i),
  Request ID (i),

>   Track Alias (i),

Reviewer: I think we should rename this at some point, because it's really an alas for the subscription, but it's ok for now.

Reviewer 2: > I think we should rename this at some point, because it's really an alas for the subscription, but it's ok for now.

I don't care if we rename it but it's not really an alias for the subscribe but a local scoped map to the full track name. Depending on other PR, publish would also use this ID. Something might make one subscribe to a track, end it, then make new subscribe to same track, Those two subscribes might on some implementaions get the same track alias for both of them because they were both to the same track.  

I fine with renaming it at some point to make it clear this is not the same thing as track alias in earlier version of the track, but I don't want to treat it like it is the same things as subscribe id.

Author: I was trying to leave open the possibility that this identifier gets bound to the track, as Cullen says.  There's also cases if we allow multiple concurrent subscriptions to the same track with non-overlapping filters.

Reviewer: I think either case can be useful.  I think "Publish ID" would be good, given it's obvious and not used elsewhere.

Author: Let's paint the possible renaming bikeshed later.

* Request ID: The Request ID of the SUBSCRIBE this message is replying to
  {{message-subscribe-req}}.

* Track Alias: The identifer used for this track in Subgroups or Datagrams (see
  {{track-alias}}). The same Track Alias MUST NOT be used to refer to two

>   different Tracks simultaneously. If a subscriber receives a SUBSCRIBE_OK that
>   uses the same Track Alias as a different track with an active subscription, it
>   MUST close the session with error 'Duplicate Track Alias'.

Reviewer: ```suggestion
  different Tracks (and therefore Subscriptions) simultaneously. If a subscriber receives a SUBSCRIBE_OK that
  uses the same Track Alias as a different track with an active subscription, it
  MUST close the session with error 'Duplicate Track Alias'.
```

Author: I'm being kind of intentionally vague here until we completely nail "multiple concurrent subscriptions".  Per last week, it seems like we're leaning "no concurrent subscriptions to same track, allow update to widen filter, publisher acknowledgement to subscribe_update".

The text I have is valid either way.

Reviewer 3: If a subscribers have multiple subscriptions to same track and if they overlap, the subscriber will indeed get objects sent multiple times for the overlapped region of the subscriptions. I think it is a fine behavior and believe that the core transport shouldn't block concurrent subscriptions to a given track.

response to a SUBSCRIBE, it SHOULD close the session with an error of 'Protocol
Violation'

> ## Track Alias

Reviewer 3: ```suggestion
## Track Alias {#track-alias}
```

Author: Through some dark magic, an anchor with the name #foo-bar is automatically created for a section titled Foo Bar.  So this isn't necessary.

To optimize wire efficiency, Subgroups and Datagrams refer to a track by a
numeric identifier, rather than the Full Track Name.  Track Alias is chosen by
the publisher and included in SUBSCRIBE_OK ({{message-subscribe-ok}}.

Though a Track Alias is only required to be unique within a single session,
Publishers MAY set Track Alias to a more unique (possibly globally unique)

> identifier for the Track.  Relays SHOULD assign the Track Alias received in an

Reviewer 2: Nope, this is total non starter for me, if we are going to have this SHOULD, we need to go back to also having the track error. This is simply not going to work. 

Imagine the case where for meetings the client all use 1 for their first track, then 2, etc. Nearly every alias will need to be changed yet putting this SHOULD here is going to make applications think they can assume t not be changed.

Author: This problem exists already with subscriber chosen alias and RETRY.  The two original publishers both selected Alias=1 on different sessions, but the downstream subscriber can only have a single track with Alias=1.

Would this work any better for you?

> Relays MUST assign the Track Alias received in SUBSCRIBE_OK to downstream subscriptions to the same Track, unless another Track is already using that Alias.

Reviewer 4: There is also a bit of a tension here with the "default case": when there are no global IDs, the relay would want to always use the lowest unused track alias for the connection.

Reviewer 2: after discussion of this ... I am OK with just going to we don't say anything about what relays do with the numbers (and of course they can not conflict)

Author: I removed the normative relay text.

> Though a Track Alias is only required to be unique within a single session,
> Publishers MAY set Track Alias to a more unique (possibly globally unique)
> identifier for the Track.  Relays SHOULD assign the Track Alias received in an
> upstream SUBSCRIBE_OK to downstream subscriptions for the same track. Since
> subscribers can request Tracks from uncoordinated publishers through a single
> relay session, it is not always possible to reuse the upstream Track Alias.
> 

Author: ```suggestion
```

0x03 and Extension Headers Length is 0, it MUST close the session with Protocol
Violation.

* Track Alias: The Track Alias ({{track-alias}} indicating the track this
  Datagram belongs to.  If an endpoint receives a datagram with an unknown Track
  Alias, it MAY drop the datagram or choose to buffer it for a brief period to

>   handle reordering with the control message that establishes the Track Alias.

Reviewer 2: Nit ... this seems like the the wrong place to put the "drop of buffer" text but I agree with the text.

Author: I sort of agree but not sure where else to put it.

Reviewer: Agreed that this is probably the wrong spot for this.  How about: https://github.com/moq-wg/moq-transport/blob/main/draft-ietf-moq-transport.md#delivery-timeout-parameter-delivery-timeout

Alternatively. we could add a section for Datagrams adjacent to Streams?

Author: I took your second suggestion and made a top level datagram section

### Subgroup Header

All Objects on a Subgroup stream belong to the track identified by `Track Alias`
(see {{track-alias}}) and the Subgroup indicated by 'Group ID' and `Subgroup
ID` in the SUBGROUP_HEADER.

If an endpoint receives a subgroup with an unknown Track Alias, it MAY abandon
the stream, or choose to buffer it for a brief period to handle reordering with
the control message that establishes the Track Alias.  The endpoint MAY withhold
stream flow control beyond the SUBGROUP_HEADER until the Track Alias has been

> established.  To prevent deadlocks, the publisher MUST allocate connection flow
> control to the control stream before allocating it any data streams.

Author: ```suggestion
established.  To prevent deadlocks, the publisher MUST allocate connection flow
control to the control stream before allocating it any data streams. Otherwise,
a receiver might wait for a control message containing a Track Alias to release
flow control, while the sender waits for flow control to send the message.
```

### Subgroup Header

All Objects on a Subgroup stream belong to the track identified by `Track Alias`
(see {{track-alias}}) and the Subgroup indicated by 'Group ID' and `Subgroup
ID` in the SUBGROUP_HEADER.

If an endpoint receives a subgroup with an unknown Track Alias, it MAY abandon
the stream, or choose to buffer it for a brief period to handle reordering with

> the control message that establishes the Track Alias.  The endpoint SHOULD NOT

Reviewer 2: I think you could make this a MAY but I strongly object to it being a SHOULD - this data arrived because the subscriber asked for this track. I see no reason not to deliver it. Doing this SHOULD will delay it getting there and make the user experience worse. Sure there is some DDOS buffer limit of what the client may stop but that has to do with DOS stuff that we can deal with across all the stuff.

Author: > I think you could make this a MAY but I strongly object to it being a SHOULD

I'm confused.  It is a MAY:

> it MAY abandon the stream, or choose to buffer it for a brief period

What are you asking for?  MUST buffer data with unknown identifiers?  I'm not sure that's any better because how much and for how long? No one has infinite memory, and it could be an attack so I don't see how we can make buffering everything forever a requirement.  If a subscriber's buffer gets filled up and it drops some stuff, it can always fetch it.  Applications that care will set buffer limits that don't impact performance.

> this data arrived because the subscriber asked for this track

Data with an unknown identifier could be due to a buggy or malicious peer.

Author: Chatted offline - I was the confused one, Cullen was referring to releasing flow control.  I'll make this a MAY.

release stream flow control beyond the SUBGROUP_HEADER until the Track Alias has

> been established.  TODO: talk about possible deadlocks.

Reviewer: Is this because the application layer can't really use this subgroup data yet?  If so, maybe say why?

Author: If the connection flow control is used up by streams with an unknown identifier it could block control messages that establish the indentifiers.

> the control message that establishes the Track Alias.  The endpoint SHOULD NOT
> release stream flow control beyond the SUBGROUP_HEADER until the Track Alias has
> been established.  TODO: talk about possible deadlocks.

Author: ```suggestion
the control message that establishes the Track Alias.  The endpoint MAY withhold
stream flow control beyond the SUBGROUP_HEADER until the Track Alias has been
established.  To prevent deadlocks, the publisher MUST allocate connection flow
control to the control stream before allocating it any data streams.
```
